homeassistant/components/upcloud/config_flow.py

```diff
@@ -0,0 +1,111 @@
+"""Config flow for UpCloud."""
+
+import logging
+
+import requests.exceptions
+import upcloud_api
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_SCAN_INTERVAL, CONF_USERNAME
+from homeassistant.core import callback
+
+# pylint: disable=unused-import # for DOMAIN https://github.com/PyCQA/pylint/issues/3202
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class UpCloudConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """UpCloud config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    username: str
+    password: str
+
+    async def async_step_user(self, user_input=None):
+        """Handle user initiated flow."""
+        if user_input is None:
+            return await self._async_show_form(step_id="user")
+
+        await self.async_set_unique_id(user_input[CONF_USERNAME])
+
+        manager = upcloud_api.CloudManager(
+            user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+        )
+
+        errors = {}
+        try:
+            await self.hass.async_add_executor_job(manager.authenticate)
+        except upcloud_api.UpCloudAPIError:
+            errors["base"] = "invalid_auth"
+            _LOGGER.debug("invalid_auth", exc_info=True)
+        except requests.exceptions.RequestException:
+            errors["base"] = "cannot_connect"
+            _LOGGER.debug("cannot_connect", exc_info=True)
+
+        if errors:
+            return await self._async_show_form(
+                step_id="user", user_input=user_input, errors=errors
+            )
+
+        return self.async_create_entry(title=user_input[CONF_USERNAME], data=user_input)
+
+    async def async_step_import(self, user_input=None):
+        """Handle import initiated flow."""
+        await self.async_set_unique_id(user_input[CONF_USERNAME])
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_user(user_input=user_input)
+
+    async def _async_show_form(self, step_id, user_input=None, errors=None):
+        """Show our form."""
+        if user_input is None:
+            user_input = {}
+        return self.async_show_form(
+            step_id=step_id,
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
```

Comment:
we shouldn't send the password back if it was wrong.

Reply:
Good point, but here we don't know if it was the username or password that was wrong (or both). If it was the username only, it should be easy for the user to visually inspect and fix, and in that case the password can be re-sent as is. If there's nothing wrong with the username, the user can just overwrite/re-enter the password -- they are in a better position to know than we are.

Reply:
Right, but it's a security measure. We should try to limit the amounts that we send a password over the wire.

Reply:
So I gather you'd rather have the user always re-enter the password, no matter why the setup failed? Including possible network errors etc.

Reply:
(By the way, it appears a bunch of integrations reuse the input default password currently, `git grep CONF_PASSWORD | grep homeassistant/ | grep default=` lists suspects and contains some false positives but also ones I believe do this.)

Reply:
Yeah, always re-enter. We should clean up the others too.